backend.native/tests/runtime/workers/worker11.kt

```diff
@@ -40,9 +39,56 @@ fun initJobs(count: Int) = Array<Job?>(count) { i -> Job(i, i * 2, i)}
             consumed++
         }
     }
-    assertEquals(consumed, COUNT)
+    assertEquals(consumed, workers.size)
     workers.forEach {
         it.requestTermination().result
     }
     println("OK")
+}
+
+val COUNT = 2
+
+@SharedImmutable
+val counters = Array(COUNT) { AtomicInt(0) }.freeze()
+
+@Test fun runTest1() {
+  val workers = Array(COUNT) { Worker.start() }
+  // Ensure processQueue() can only be called on current Worker.
+  workers.forEach {
+      assertFailsWith<IllegalStateException> {
+          it.processQueue()
+      }
+  }
+  val futures = Array(workers.size) { workerIndex ->
+      workers[workerIndex].execute(TransferMode.SAFE, {
+          workerIndex
+      }) {
+          index ->
+          assertEquals(0, counters[index].value)
+          // Process following request.
+          while (!Worker.current!!.processQueue()) {}
+          // Ensure it has an effect.
+          assertEquals(1, counters[index].value)
+          // No more non-terminating tasks in this worker queue.
+          assertEquals(false, Worker.current!!.processQueue())
+      }
+  }
+  val futures2 = Array(workers.size) { workerIndex ->
+      workers[workerIndex].execute(TransferMode.SAFE, {
+          workerIndex
+      }) { index ->
+          assertEquals(0, counters[index].value)
+          counters[index].increment()
+      }
+  }
+  futures2.forEach { it.result }
+  futures.forEach { it.result }
+  workers.forEach {
+      it.requestTermination().result
+  }
+
+  // Ensure terminated workers are no longer there.
+  workers.forEach {
+    assertFailsWith<IllegalStateException> { it.execute(TransferMode.SAFE, { Unit }) { println("ERROR") } }
```

Comment:
Btw, is the implementation of this behaviour deterministic?

https://github.com/JetBrains/kotlin-native/blob/7811831674563f02b21409b9fa06b3dab7dd73f3/runtime/src/main/cpp/Worker.cpp#L540
```c++
job.terminationRequest.future->storeResultUnlocked(nullptr, true);
theState()->removeWorkerUnlocked(id());
break;
```

I.e. the future is notified before the worker is removed, so it is not guaranteed that `it.execute` throws an exception here.

Also what happens if the job is added after termination request was sent but before it gets completed?

Reply:
Swapped removing and notification statements. 

If job is added after termination request is sent, it depends how adding time correlates with the termination handler and if handler shall wait. Generally, if we do processing of pending requests -  we process whatever is available in the queue at the moment of processing. If we'll terminate unconditionally - queue content is silently ignored.